Getting started doc content (#178)
microsoft/reverse-proxy: docs/docfx/articles/getting_started.md

---
uid: getting_started
title: Getting Started with YARP
---

# Getting Started with YARP

YARP is designed as a library that provides the core proxy functionality which you can then customize by adding or replacing modules. YARP is currently supplied as a nuget package and code snippets. We plan on having a project template, and pre-built exe in the future. 

YARP supports ASP.NET Core 3.1 or 5.0 preview4 or later. You can download preview 4 of .NET 5 SDK from https://dotnet.microsoft.com/download/dotnet/5.0.

### Create a new project

Start by creating an "Empty" ASP.NET Core application using the command line:

```
dotnet new web -n MyProxy 
```

use `-f` to specify `netcoreapp3.1` or `netcoreapp5.0`.

Or create a new ASP.NET Core web application in Visual Studio, and choose "Empty" for the project template. 

### Update the project file

Open the Project and make sure it includes the appropriate target framework: 
 
 ```
<PropertyGroup>

>   <TargetFramework>netcoreapp5.0</TargetFramework>

Reviewer: nit: net5.0

Author: The preview4 templates still use netcoreapp5.0.

---
uid: getting_started
title: Getting Started with YARP
---

# Getting Started with YARP

> YARP is designed as a library that provides the core proxy functionality which you can then customize by adding or replacing modules. YARP is currently supplied as a nuget package and code snippets. We plan on having a project template, and pre-built exe in the future. 

Reviewer 2: Nit:

```suggestion
YARP is designed as a library that provides the core proxy functionality which you can then customize by adding or replacing modules. YARP is currently provided as a NuGet package and code snippets. We plan on providing a project template and pre-built exe in the future. 
```

> YARP supports ASP.NET Core 3.1 or 5.0 preview4 or later. You can download preview 4 of .NET 5 SDK from https://dotnet.microsoft.com/download/dotnet/5.0.

Reviewer 2: Is and better than or here?

```suggestion
YARP supports ASP.NET Core 3.1 and 5.0.0-preview.4 or later. You can download preview 4 of .NET 5 SDK from https://dotnet.microsoft.com/download/dotnet/5.0.
```

Author: I've applied all of the feedback except the 5.0.0-preview.4 name change. 5.0.0-preview.4 is the format used in file names, not the name used when discussing the release. 
https://devblogs.microsoft.com/aspnet/asp-net-core-updates-in-net-5-preview-4/

Start by creating an "Empty" ASP.NET Core application using the command line:

```
dotnet new web -n MyProxy 
```

use `-f` to specify `netcoreapp3.1` or `netcoreapp5.0`.

Or create a new ASP.NET Core web application in Visual Studio, and choose "Empty" for the project template. 

### Update the project file

Open the Project and make sure it includes the appropriate target framework: 
 
 ```
<PropertyGroup>
  <TargetFramework>netcoreapp5.0</TargetFramework>
</PropertyGroup> 
```

And then add:
 
 ```
<ItemGroup> 
  <PackageReference Include="Microsoft.ReverseProxy" Version="1.0.0-preview.1.*" /> 
</ItemGroup> 
```

### Update Startup

YARP is implemented as a ASP.NET Core component, and so the majority of the sample code is in Startup.cs. 

YARP currently uses configuration files to define the routes and endpoints for the proxy. That is loaded in the `ConfigureServices` method. 

```
public IConfiguration Configuration { get; }

public Startup(IConfiguration configuration)
{
    Configuration = configuration;
}

public void ConfigureServices(IServiceCollection services) 
{ 
    services.AddReverseProxy() 
        .LoadFromConfig(Configuration.GetSection("ReverseProxy")); 
} 
```

> The `Configure` method defines the ASP.NET pipeline for processing requests. The reverse proxy is plugged in to ASP.NET endpoint routing, and then has its own sub pipeline for the proxy. Here proxy pipeline modules, such as load balancing can be added to customize the handling of the request. 

Reviewer 2: Nit:

```suggestion
The `Configure` method defines the ASP.NET pipeline for processing requests. The reverse proxy is plugged into ASP.NET endpoint routing and then has its own sub-pipeline for the proxy. Here proxy pipeline modules, such as load balancing, can be added to customize the handling of the request. 
```

```
public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
{
    if (env.IsDevelopment())
    {
        app.UseDeveloperExceptionPage();
    }

    app.UseRouting();
    app.UseEndpoints(endpoints => 
    {
        endpoints.MapReverseProxy(proxyPipeline => 
        { 
            proxyPipeline.UseProxyLoadBalancing(); 
        }); 
    }); 
} 
```
 
### Configuration 

The configuration for YARP is defined in the appsettings.json file. It defines a set of:

Routes - which map incoming requests to the backend clusters based on aspects of the request such as host name, path, method, request headers etc. Specifying a Host is the only required field. Routes are ordered, so the "app1" route needs to be defined first as "route2" will act as a catchall for all paths that have not already been matched. 

Backends - which are the clusters of destination servers that requests can be routed to and load balanced across.

> Address is the URI prefix that will have the original request path appended to it.

Reviewer 2: Should we also call out the query string here?

```suggestion
Address is the URI prefix that will have the original request path and query string appended to it.
```
